src/build_helper/src/git.rs

```diff
@@ -27,145 +28,234 @@ pub fn output_result(cmd: &mut Command) -> Result<String, String> {
+/// Returns the latest upstream commit that modified `target_paths`, or `None` if no such commit
+/// was found.
+fn get_latest_upstream_commit_that_modified_files(
+    git_dir: &Path,
+    git_config: &GitConfig<'_>,
+    target_paths: &[&str],
+) -> Result<Option<String>, String> {
     let mut git = Command::new("git");
-    if let Some(git_dir) = git_dir {
-        git.current_dir(git_dir);
+    git.current_dir(git_dir);
+
+    // In theory, we could just use
+    // `git rev-list --first-parent HEAD --author=<merge-bot> -- <paths>`
+    // to find the latest upstream commit that modified `<paths>`.
+    // However, this does not work if you are in a subtree sync branch that contains merge commits
+    // which have the subtree history as their first parent, and the rustc history as second parent:
+    // `--first-parent` will just walk up the subtree history and never see a single rustc commit.
+    // We thus have to take a two-pronged approach. First lookup the most recent upstream commit
+    // by *date* (this should work even in a subtree sync branch), and then start the lookup for
+    // modified paths starting from that commit.
+    //
+    // See https://github.com/rust-lang/rust/pull/138591#discussion_r2037081858 for more details.
+    let upstream = get_closest_upstream_commit(Some(git_dir), git_config, CiEnv::None)?
+        .unwrap_or_else(|| "HEAD".to_string());
+
+    git.args([
+        "rev-list",
+        "--first-parent",
+        "-n1",
+        &upstream,
+        "--author",
+        git_config.git_merge_commit_email,
+    ]);
+
+    if !target_paths.is_empty() {
+        git.arg("--").args(target_paths);
     }
-    Ok(output_result(git.arg("merge-base").arg(&updated_master).arg("HEAD"))?.trim().to_owned())
+    let output = output_result(&mut git)?.trim().to_owned();
+    if output.is_empty() { Ok(None) } else { Ok(Some(output)) }
 }
 
-/// Searches for the nearest merge commit in the repository that also exists upstream.
+/// Returns the most recent (ordered chronologically) commit found in the local history that
+/// should exist upstream. We identify upstream commits by the e-mail of the commit
+/// author.
 ///
-/// It looks for the most recent commit made by the merge bot by matching the author's email
-/// address with the merge bot's email.
-pub fn get_closest_merge_commit(
+/// If we are in CI, we simply return our first parent.
+fn get_closest_upstream_commit(
     git_dir: Option<&Path>,
     config: &GitConfig<'_>,
-    target_paths: &[&str],
-) -> Result<String, String> {
+    env: CiEnv,
+) -> Result<Option<String>, String> {
+    let base = match env {
+        CiEnv::None => "HEAD",
+        CiEnv::GitHubActions => {
+            // On CI, we should always have a non-upstream merge commit at the tip,
+            // and our first parent should be the most recently merged upstream commit.
+            // We thus simply return our first parent.
```

Comment:
Is the logic below ever actually wrong? i.e., is this an optimization or a critical condition?

Reply:
In theory it could be wrong if the PR contained a bors commit that has a newer date than the latest merged bors commit. But this should not happen, because subtrees no longer use bors, and we shouldn't be merging any other bors merge commits in the PR (definitely not newer ones than the latest upstream), unless someone tries to do something funky.